File: development/C++/samples/server/DemoServer/Transaction.cpp
#include "stdafx.h"
#include "Transaction.h"
#include "Function.h"

Transaction::Transaction(
	EnumTransactionType aTransactionType,
	std::vector<DaRequest*>& aRequestList,
	unsigned long aSessionKey) :
	DaTransaction(
		aTransactionType,
		aRequestList,
		aSessionKey)
{
}   //  end constructor

Transaction::~Transaction()
{
}   //  end destructor

long Transaction::handleReadRequests(void)
{
	size_t count = m_requestList.size();
	DaRequest *request = NULL;

	for (size_t i = 0; i < count; i++)
	{
		request = m_requestList[i];

		if ((request != NULL) && (request->progressRequestState(EnumRequestState_CREATED, EnumRequestState_INPROGRESS) == true))
		{
			DemoDaAddressSpaceElement* pElement = (DemoDaAddressSpaceElement*)request->getAddressSpaceElement();

			if (pElement == NULL)
			{
				request->setResult(E_FAIL);
			}
			else
			{
				pElement->handleReadRequest(request);
			}   //  end if ... else
		}   //  end if
	}   //  end for

	return completeRequests();
}   //  end HandleReadRequests

long Transaction::handleWriteRequests(void)
{
	size_t count = m_requestList.size();
	DaRequest *request = NULL;

	for (size_t i = 0; i < count; i++)
	{
		request = m_requestList[i];

		if ((request != NULL) && (request->progressRequestState(EnumRequestState_CREATED, EnumRequestState_INPROGRESS) == true))
		{
			DemoDaAddressSpaceElement* pElement = (DemoDaAddressSpaceElement*)request->getAddressSpaceElement();

			if (pElement == NULL)
			{
				m_requestList[i]->setResult(E_FAIL);
			}
			else
			{
				pElement->handleWriteRequest(m_requestList[i]);
			}   //  end if ... else
		}   //  end if
	}   //  end  for

	return completeRequests();
}   //  end handleWriteRequests

File: development/NET/samples/server/DemoServer/CS/Transaction.cs
using Softing.OPCToolbox;
using Softing.OPCToolbox.Server;

namespace DemoServer
{
	public class Transaction : DaTransaction
	{
		#region Public Methods
		//--------------------

		public Transaction(
			EnumTransactionType aTransactionType,
			DaRequest[] requestList,
			uint aSessionKey)
			:
			base(
			aTransactionType,
			requestList,
			aSessionKey)
		{
		}   //	end constructor

		public override int HandleReadRequests()
		{
			int count = Requests.Count;
			for (int i = 0; i < count; i++)
			{
				DaRequest request = Requests[i] as DaRequest;
				if ((request != null) && (request.ProgressRequestState(EnumRequestState.CREATED, EnumRequestState.INPROGRESS) == true))
				{
					DemoDaAddressSpaceElement element = (DemoDaAddressSpaceElement)request.AddressSpaceElement;
					if (element != null)
					{
						element.HandleReadRequest(request);
					}
					else
					{
						request.Result = EnumResultCode.E_FAIL;
					}
				}   //	end if
			}   //	end for
			return CompleteRequests();
		}   //	end HandleReadRequests

		public override int HandleWriteRequests()
		{
			int count = Requests.Count;
			for (int i = 0; i < count; i++)
			{
				DaRequest request = Requests[i] as DaRequest;
				if ((request != null) && (request.ProgressRequestState(EnumRequestState.CREATED, EnumRequestState.INPROGRESS) == true))
				{
					DemoDaAddressSpaceElement element = (DemoDaAddressSpaceElement)request.AddressSpaceElement;
					if (element != null)
					{
						element.HandleWriteRequest(request);
					}
					else
					{
						request.Result = EnumResultCode.E_FAIL;
					}
				}   //	end if
			}   //	end for
			return CompleteRequests();
		}   //	end HandleWriteRequests

		//--
		#endregion
	}   //	end class MyTransaction
}   //	end namespace
